### PR TITLE
Dropout

### PR DESCRIPTION
**Pull request by szagoruyko**
cudnn.Dropout should take about a ton less memory because it stores masks in binary format and faster as well, thanks to @ngimel for explaining this.
can someone review please?

Btw, R5 is released, it's time to make it master @soumith ?

### CODE REVIEW

File: Dropout.lua [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+   if not self.dropDesc then
+      self.dropDesc = ffi.new('struct cudnnDropoutStruct*[1]')
+      errcheck('cudnnCreateDropoutDescriptor', self.dropDesc)
+      local reserves_size = getSize('cudnnDropoutGetReserveSpaceSize', self.iDesc[0])
+      self.reserves = self.reserves or torch.CudaByteTensor()
+      self.reserves = self.reserves:cudaByte():resize(reserves_size)
+      local state = cudnn.dropout_rng_states[dev]
+      errcheck('cudnnSetDropoutDescriptor', self.dropDesc[0],
+         cudnn.getHandle(), self.p,
+         state:data(), state:nElement(), torch.seed())
```
Comment (ngimel) [inline Dropout.lua]: this would reset the seed of torch RNG - this is probably not something you want to do, as it can introduce non-desirable non-determinism. Better to use torch.random() here?

**Comment by ngimel** [comment]
cudnn.Dropout:setp() will be broken (for it to work, cudnn.Dropout:setp() should call setDropoutDescriptor using NULL as state argument if states are initialized already).

**Comment by ngimel** [comment]
There is a potential problem with several dropout layers sharing the states and running concurrently in the different streams, but its hard to make this scenario completely safe.
